File: implementations/python/src/upp/models/enums.py
"""UPP Enumerations.

Defines the enumeration types used throughout the Universal Personalization
Protocol. All enums inherit from ``str`` and ``enum.Enum`` so they
serialize naturally to their string values in JSON.

Enumerations:
    EventStatus: Lifecycle state of a stored event.
    SourceType: Provenance classification of an extracted fact.
    SensitivityTier: Privacy classification for ontology labels.
    Cardinality: Whether a label accepts one or many values.
    Durability: Expected lifespan of a fact.
"""

from __future__ import annotations

import enum

__all__ = [
    "Cardinality",
    "Durability",
    "EventStatus",
    "SensitivityTier",
    "SourceType",
    "TaskStatus",
]


class EventStatus(enum.StrEnum):
    """Lifecycle state of a stored event.

    Events follow an immutable event-sourcing pattern:

    - ``VALID`` — Active and authoritative.
    - ``STAGED`` — Low confidence, pending reinforcement.
    - ``SUPERSEDED`` — Replaced by a newer event, retained for audit.
    """

    VALID = "valid"
    """Current, retrievable event."""

    STAGED = "staged"
    """Low-confidence event awaiting reinforcement."""

    SUPERSEDED = "superseded"
    """Replaced by a newer event. Retained for audit trail."""


class SourceType(enum.StrEnum):
    """Classifies how a fact was obtained.

    Source types indicate the provenance of an extracted fact.
    """

    USER_STATED = "user_stated"
    """The user explicitly stated this fact."""

    AGENT_OBSERVED = "agent_observed"
    """The agent observed or derived this fact from conversation context."""

    INFERRED = "inferred"
    """The system inferred this fact from indirect signals."""


class SensitivityTier(enum.StrEnum):
    """Privacy classification for ontology labels.

    Ordered from least to most sensitive:
    ``TIER_PUBLIC < TIER_WORK < TIER_PERSONAL < TIER_SENSITIVE < TIER_INTERNAL``
    """

    TIER_PUBLIC = "tier_public"
    """Safe to share broadly."""

    TIER_WORK = "tier_work"
    """Professional context."""

    TIER_PERSONAL = "tier_personal"
    """Personal but non-sensitive."""

    TIER_SENSITIVE = "tier_sensitive"
    """Sensitive personal data."""

    TIER_INTERNAL = "tier_internal"
    """Never shared externally."""


class Cardinality(enum.StrEnum):
    """Defines whether a label accepts one or many concurrent values.

    Cardinality determines supersession behavior: singular labels cause
    new events to supersede old ones, while plural labels accumulate.
    """

    SINGULAR = "singular"
    """Only one value at a time. New values supersede old ones."""

    PLURAL = "plural"
    """Multiple values coexist. New values are added alongside existing ones."""


class Durability(enum.StrEnum):
    """Indicates how long a fact is expected to remain valid.

    Ordering (longest to shortest): ``PERMANENT > TRANSIENT > EPHEMERAL``
    """

    PERMANENT = "permanent"
    """Unlikely to change over a person's lifetime."""

    TRANSIENT = "transient"
    """Changes over months or years."""

    EPHEMERAL = "ephemeral"
    """Changes frequently — days or weeks."""


class TaskStatus(enum.StrEnum):
    """Status of a background task."""

    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"

File: implementations/typescript/src/models/enums.ts
/**
 * @file enums.ts
 * @description UPP protocol enumerations with Zod validation schemas.
 *
 * Defines all enumeration types used throughout the Universal Personalization
 * Protocol. Each enum is represented as a Zod schema for validation and an
 * inferred TypeScript type for compile-time safety.
 *
 * Enumerations:
 *   EventStatus — Lifecycle state of a stored event.
 *   SourceType — Provenance classification of an extracted fact.
 *   SensitivityTier — Privacy classification for ontology labels.
 *   Cardinality — Whether a label accepts one or many values.
 *   Durability — Expected lifespan of a fact.
 *
 * @module
 */

import { z } from "zod";

// ─────────────────────────────────────────────────────────────────────────────
// EventStatus
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Zod schema for event lifecycle status.
 *
 * Events follow an immutable event-sourcing pattern:
 * - `valid` — Active and authoritative.
 * - `staged` — Low confidence, pending reinforcement.
 * - `superseded` — Replaced by a newer event, retained for audit.
 */
export const EventStatusSchema = z.enum(["valid", "staged", "superseded"]);

/** Lifecycle state of a stored event. */
export type EventStatus = z.infer<typeof EventStatusSchema>;

/** All valid EventStatus values as a readonly array. */
export const EVENT_STATUS_VALUES = EventStatusSchema.options;

// ─────────────────────────────────────────────────────────────────────────────
// SourceType
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Zod schema for fact provenance classification.
 *
 * Source types indicate how a fact was obtained:
 * - `user_stated` — The user explicitly stated this fact.
 * - `agent_observed` — The agent observed or derived this from conversation.
 * - `inferred` — The system inferred this from indirect signals.
 */
export const SourceTypeSchema = z.enum(["user_stated", "agent_observed", "inferred"]);

/** Provenance classification of an extracted fact. */
export type SourceType = z.infer<typeof SourceTypeSchema>;

/** All valid SourceType values as a readonly array. */
export const SOURCE_TYPE_VALUES = SourceTypeSchema.options;

// ─────────────────────────────────────────────────────────────────────────────
// SensitivityTier
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Zod schema for privacy sensitivity classification.
 *
 * Ordered from least to most sensitive:
 * `tier_public < tier_work < tier_personal < tier_sensitive < tier_internal`
 */
export const SensitivityTierSchema = z.enum([
  "tier_public",
  "tier_work",
  "tier_personal",
  "tier_sensitive",
  "tier_internal",
]);

/** Privacy classification for ontology labels. */
export type SensitivityTier = z.infer<typeof SensitivityTierSchema>;

/** All valid SensitivityTier values as a readonly array. */
export const SENSITIVITY_TIER_VALUES = SensitivityTierSchema.options;

// ─────────────────────────────────────────────────────────────────────────────
// Cardinality
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Zod schema for label cardinality.
 *
 * Determines supersession behavior:
 * - `singular` — Only one value at a time. New values supersede old ones.
 * - `plural` — Multiple values coexist. New values are added alongside.
 */
export const CardinalitySchema = z.enum(["singular", "plural"]);

/** Whether a label accepts one or many concurrent values. */
export type Cardinality = z.infer<typeof CardinalitySchema>;

/** All valid Cardinality values as a readonly array. */
export const CARDINALITY_VALUES = CardinalitySchema.options;

// ─────────────────────────────────────────────────────────────────────────────
// Durability
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Zod schema for fact durability.
 *
 * Indicates how long a fact is expected to remain valid:
 * - `permanent` — Unlikely to change over a person's lifetime.
 * - `transient` — Changes over months or years.
 * - `ephemeral` — Changes frequently — days or weeks.
 */
export const DurabilitySchema = z.enum(["permanent", "transient", "ephemeral"]);

/** Expected lifespan of a fact. */
export type Durability = z.infer<typeof DurabilitySchema>;

/** All valid Durability values as a readonly array. */
export const DURABILITY_VALUES = DurabilitySchema.options;

// ─────────────────────────────────────────────────────────────────────────────
// TaskStatus
// ─────────────────────────────────────────────────────────────────────────────

/** Status of a background task. */
export const TaskStatusValues = ["pending", "running", "completed", "failed"] as const;

/** Zod schema for TaskStatus. */
export const TaskStatusSchema = z.enum(TaskStatusValues);

/** TypeScript type for task status. */
export type TaskStatus = z.infer<typeof TaskStatusSchema>;
